target/linux/ramips/image/mt7621.mk

```diff
@@ -289,7 +289,8 @@ define Device/asus_rt-ax53u
   IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
   IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | append-ubi | \
 	check-size
-  DEVICE_PACKAGES := kmod-mt7915e kmod-usb3 uboot-envtools
+  DEVICE_PACKAGES := kmod-mt7915e kmod-usb3 kmod-usb-ledtrig-usbport \
+	uboot-envtools
```

Comment:
To make use of the uboot-envtools you have to add support for this device to package/boot/uboot-envtools/files/ramips

Reply:
How to test that it work after I made changes to ```package/boot/uboot-envtools/files/ramips```?

Reply:
check if `fw_printenv` and `fw_setenv` are working. They should read and write to the U-Boot environment.

Reply:
Even though the original firmware uses a partition scheme where an `nvram` partition exists from `0x0e0000` to `0x1e0000`, and this partition is also referred to as `u-boot-env` by the u-boot environment variables, they are not stored there. The u-boot environment variables are actually stored at `0x80000` which is inside the `u-boot` partition.

So the simplest way to make uboot-envtools work would be to add:
```sh
asus,rt-ax53u)
	ubootenv_add_uci_config "/dev/mtd0" "0x80000" "0x20000" "0x20000" "3"
	;;
```
The size for the env vars is `0x20000`, which is also the NAND sector size, and there is space for 3 env var blocks within the remaining space of the `u-boot` partiton. For `fw_setenv` to work the `u-boot` partition must not be set to `read-only`, otherwise writing to it is impossible.

Another way of solving this would be to split the `u-boot` partition at `0x80000`, leaving the `u-boot` partition as `read-only` and making the `u-boot-env` partition start at `0x80000` instead of `0xe0000`.

Reply:
I think split the u-boot partition is a good way.
```
		partition@0 {
			label = "u-boot";
			reg = <0x0 0x80000>;
			read-only;
		};

		/*
		 * In stock firmware, "u-boot" partition include this "u-boot-env" partition.
		 * We split it to prevent unexpected writing to "u-boot" partition.
		 */
		partition@80000 {
			label = "u-boot-env";
			reg = <0x80000 0x60000>;
		};

		partition@e0000 {
			label = "nvram";
			reg = <0xe0000 0x100000>;
			read-only;
		};
```

Reply:
I just tried out @DragonBluep's solution and it doesn't appear to work.
```
root@OpenWrt:~# fw_printenv autoload
Cannot parse config file '/etc/fw_env.config': No such file or directory
Failed to find NVMEM device
root@OpenWrt:~# fw_printenv autostart
Cannot parse config file '/etc/fw_env.config': No such file or directory
Failed to find NVMEM device
root@OpenWrt:~# fw_printenv
Cannot parse config file '/etc/fw_env.config': No such file or directory
Failed to find NVMEM device
```

in-case I made a mistake, these are the mtd partitions:
logread
```
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.661935] 8 fixed-partitions partitions found on MTD device mt7621-nand
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.669550] Creating 8 MTD partitions on "mt7621-nand":
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.674780] 0x000000000000-0x000000080000 : "u-boot"
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.688732] 0x000000080000-0x0000000e0000 : "u-boot-env"
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.700748] 0x0000000e0000-0x0000001e0000 : "nvram"
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.722163] 0x0000001e0000-0x0000002e0000 : "factory"
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.743719] 0x0000002e0000-0x0000003e0000 : "factory2"
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.765504] 0x0000003e0000-0x0000007e0000 : "kernel"
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    0.833185] 0x0000007e0000-0x0000035e0000 : "ubi"
Fri Dec 23 03:52:24 2022 kern.notice kernel: [    1.543890] 0x0000035e0000-0x0000067e0000 : "firmware2"
```
cat /proc/mtd
```
root@OpenWrt:/etc# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00080000 00020000 "u-boot"
mtd1: 00060000 00020000 "u-boot-env"
mtd2: 00100000 00020000 "nvram"
mtd3: 00100000 00020000 "factory"
mtd4: 00100000 00020000 "factory2"
mtd5: 00400000 00020000 "kernel"
mtd6: 02e00000 00020000 "ubi"
mtd7: 03200000 00020000 "firmware2"
```

Reply:
@BlackMark29A's suggestion appears to work fine:
```
root@OpenWrt:~# fw_printenv
Warning: Bad CRC, using default environment
bootcmd=bootp; setenv bootargs root=/dev/nfs nfsroot=${serverip}:${rootpath} ip=${ipaddr}:${serverip}:${gatewayip}:${netmask}:${hostname}::off; bootm
bootdelay=5
baudrate=115200
```
But the warning makes me apprehensive about testing the write functionality with `fw_setenv bootdelay 6`
Was it unable to load the actual config due to a bad crc? or is the actual uboot environment of the device this small?

Reply:
If you get `Cannot parse config file '/etc/fw_env.config': No such file or directory` then your changes to `package/boot/uboot-envtools/files/ramips` weren't included in the build. You can also check to see if `/rom/etc/uci-defaults/30_uboot-envtools` contains the changes you made.

For `fw_printenv` to work you need to have a `/etc/fw_env.config` file containing an entry somehow pointing to `0x80000`, because that's where the uboot environment variables are. So depending on the partition layout this config needs to be different.

Here's an example that I tested and that definitely works:

Partitions:

```
partition@0 {
	label = "u-boot";
	reg = <0x0 0x80000>;
	read-only;
};

partition@80000 {
	label = "u-boot-env";
	reg = <0x80000 0x20000>;
};

partition@a0000 {
	label = "nvram";
	reg = <0xa0000 0x140000>;
};
```

uboot-envtools:
```
asus,rt-ax53u)
	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x20000" "0x20000"
	;;
```

This will result in the following partition layout:

```
[    2.660250] 7 fixed-partitions partitions found on MTD device mt7621-nand
[    2.667029] Creating 7 MTD partitions on "mt7621-nand":
[    2.672253] 0x000000000000-0x000000080000 : "u-boot"
[    2.678225] 0x000000080000-0x0000000a0000 : "u-boot-env"
[    2.684808] 0x0000000a0000-0x0000001e0000 : "nvram"
[    2.690635] 0x0000001e0000-0x0000002e0000 : "factory"
[    2.696757] 0x0000002e0000-0x0000003e0000 : "factory2"
[    2.702894] 0x0000003e0000-0x0000007e0000 : "kernel"
[    2.708798] 0x0000007e0000-0x000007800000 : "ubi"
```
(Ignore the fact that `ubi` is different and `firmware2` is missing. I combined them, but that has nothing to do with the u-boot partiton layout.)

and this fw_env.config file:
```
/dev/mtd1 0x0 0x20000 0x20000
```

and then `fw_printenv` will work:
```console
root@OpenWrt:~# fw_printenv
baudrate=115200
bootcmd=mtkautoboot
bootdelay=0
bootmenu_0=Startup system (Default)=tftpd
bootmenu_1=Upgrade firmware=mtkupgrade fw
bootmenu_2=Upgrade bootloader=mtkupgrade bl
bootmenu_3=Upgrade bootloader (advanced mode)=mtkupgrade bladv
bootmenu_4=Load image=mtkload
fdtcontroladdr=8ffe3600
ipaddr=192.168.1.1
mtddevname=u-boot
mtddevnum=0
mtdids=nmbm0=nmbm0
mtdparts=mtdparts=nmbm0:896k(u-boot),1024k(u-boot-env),1024k(factory),1024k(factory2),-(firmware)
netmask=255.255.255.0
partition=nmbm0,0
serverip=192.168.1.2
stderr=uartlite0@1e000c00
stdin=uartlite0@1e000c00
stdout=uartlite0@1e000c00
```

Reply:
Thanks, for helping me out :)
I actually rebuild the rom with `make clean` to make sure this time. (If `make dirclean` is necessary, then say so, but from what I read, it isn't)

1. Before that I checked both files though, and they show the exact changes that should've been required
![grafik](https://user-images.githubusercontent.com/5685512/209411362-debba7ac-8357-4853-b124-2d21aaebbc58.png)

2. With the rebuild rom and still 3 sectors:
```
root@OpenWrt:~# fw_printenv
Warning: Bad CRC, using default environment
bootcmd=bootp; setenv bootargs root=/dev/nfs nfsroot=${serverip}:${rootpath} ip=${ipaddr}:${serverip}:${gatewayip}:${netmask}:${hostname}::off; bootm
bootdelay=5
baudrate=115200
```

3. Rebuild with 1 sector and the split partition. But with the difference, that I kept nvram `read-only` (for testing whether that matters):
![grafik](https://user-images.githubusercontent.com/5685512/209416143-d35a7c04-4ba2-4003-b465-15eec57e35b6.png)
Interestingly enough, that didn't work either (in-case it matters: I am always doing sysupgrade -n)

4. Rebuild with 1 sector and the split partition and exactly the way you did it: without read-only nvram
![grafik](https://user-images.githubusercontent.com/5685512/209420490-b3f5cd83-f100-4d7b-b7e2-71302516a6a5.png)

I'm almost out of ideas. The only thing that I can think of: I never connected to my Asus over serial, so maybe the CRC never got set, because I never read env variables nor wrote to them. What do you think?

### Addendum:
> (Ignore the fact that ubi is different and firmware2 is missing. I combined them, but that has nothing to do with the u-boot partiton layout.)

Thanks for the headsup, I actually know about that other PR (10685). I was going to try that next , once I was done testing uboot-envtools :)

Reply:
So, I just double checked on a different RT-AX53U, one where I haven't experimented at all and just flashed the firmware, and I can reproduce the behavior you are seeing.

So the issue is that by default the whole u-boot-env block is filled with `0xFF`. Only once you enter u-boot and issue a `saveenv` command will the default configuration be saved to flash. And then you can read that with `fw_printenv` in openwrt.

You'd have to connect to the serial port of the router to be able to enter u-boot and issue that command, but then `fw_printenv` should work as expected.

Reply:
thanks for verifying this :)
So by default uboot checks the flash, the flash fails and it loads it's own factory defaults?

and if anyone here would provide me a dd copy of ubootenv I could theoretically mtd-write it and after a reboot my bootloader would load that config, because it has a valid crc.

I got some questions and ideas now :)
Would the bootloader read 1 sector of ubootenv or actually up to 3 sectors of variables? (do we need to test this to get the correct `fw_env.config`?)

Shall we keep nvram read-only?
Does it serve any purpose?

Theoretically it should be possible to create `fw_resetenv` by setting the partition to 0xFF or atleast the CRC position, right?
`fw_initenv` would be cool, if it is possible to get the default config from mtd0. (emulate saveenv)

Reply:
I'm not sure what exactly u-boot does, but I assume it is pretty much what you said.

Here's a dump of the default u-boot-env you can flash: [u-boot-env.img.gz](https://github.com/openwrt/openwrt/files/10300071/u-boot-env.img.gz)

I'm pretty sure the u-boot config only reads 1 sector, I vaguely remember checking that in the source code that Asus provides, but I might be misremembering.

`fw_resetenv` could easily be implemented by just erasing `mtd1`, yes. But uboot-utils doesn't have this feature, and if it was to be added it would have to work for all devices, not just that router.
Same goes for `fw_initenv`, of course. And on top of that `fw_initenv` would be tricky, because there's no copy of the default u-boot config in flash anywhere, I checked. The default config gets built by u-boot at runtime in memory, so emulating `saveenv` wouldn't be as easy as just copying something from `mtd0`.

Reply:
Thank you. I'll test it in the new year, once I have access to the device again.
I suggest that I'll create a separate PR for uboot-envtools from what we discussed here, ok?

Reply:
> Shall we keep nvram read-only?
> Does it serve any purpose?

I missed that previously, so let me answer that as well. `nvram` is used by the stock firmware to store configuration data. It's not used by openwrt at all. I wouldn't make it read-only though, as read-only (NV)RAM would be extremely silly, and writing to it causes no harm, as it's not used anyway.

A new PR would probably make sense, as this PR seems quite stale at this point.

Reply:
Since we don't use nvram at all, I feel like we should maybe leave it as is to avoid confusion, in-case people want to read/write backups from/to it, what do you think?

Reply:
With "the same" you mean having it start at `0xe0000` like in the stock layout, instead of `0x`a0000` like I did above? Yes, absolutely, my layout was not meant as a suggestion, just that I tested it with that.